GDB automation with Python script to measure memory usage in dashd

gladcow opened the pull request:
This  code allows to calc the memmory usage of almost all objects in dashd code,  attaching GDB debuger to running dashd process.

PaulieD commented:
Any reason why this is still sitting around here, with no action since May 11th? @UdjinM6 
To me this almost seems like something that should be handled externally, by the person who needs to use this feature. But at the same time I see little reason not to merge it. Thoughts?

gladcow commented:
This PR is out of date a little. It allows complex debug automation for dashd with python (to calc exact memory usage for given dashd process for now, but in the future it can be used in integration python tests to check inner node state with GDB,  for example). I can update it if you think it is good idea or modify it to create new PR with some other functionality which is more useful at the moment.

UdjinM6 commented:
> This PR is out of date a little.

You mean it references stuff like `CMasternodeIndex` which are no longer a thing? Or is there a bigger issue?

> with some other functionality which is more useful at the moment

If it's still about memory measurement and doesn't touch the cpp part too much, I guess it would be fine to continue here. If it's however touching cpp code, I would rather fix and merge this PR first and then build new functionality on top.

gladcow commented:
I've removed unused classes from the simple class list. I've failed to generate this list automatically, I'll try to add this generation if we will use this code in the future. I've rebased it to current `develop`. Re-review,  please.

gladcow commented:
Sorry, I've removed the commits from other branch,  re-review,  please

UdjinM6 commented on contrib/auto_gdb/README.md:
```diff
@@ -0,0 +1,46 @@
+This directory contains tools to automatically get data about the memory consumption by some objects in dashd process with the help of GDB debugger.
+
+## dash_dbg.sh
+This shell script attaches GDB to running dashd process (should be builded with debug info), executes debug.gdb script and detaches.
```
s/builded/built/

UdjinM6 commented on contrib/auto_gdb/README.md:
```diff
@@ -0,0 +1,46 @@
+## common_helpers.py
+Several helper functions that are used in other python code.
+
+
```
drop one empty line here pls

UdjinM6 commented on contrib/auto_gdb/dash_dbg.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+# use testnet settings,  if you need mainnet,  use ~/.dashcore/dashd.pid file instead
+dash_pid=$(<~/.dashcore/testnet3/dashd.pid)
+sudo gdb -batch -ex "source debug.gdb" dashd ${dash_pid}
```
add new line at eof pls

UdjinM6 commented on contrib/auto_gdb/used_size.py:
```diff
@@ -0,0 +1,45 @@
+            raise e
+
+UsedSizeCommand()
+
```
drop one empty line here pls

gladcow commented:
Fixed

PastaPastaPasta commented on contrib/auto_gdb/README.md:
```diff
@@ -0,0 +1,45 @@
+This directory contains tools to automatically get data about the memory consumption by some objects in dashd process with the help of GDB debugger.
+
+## dash_dbg.sh
+This shell script attaches GDB to running dashd process (should be built with debug info), executes debug.gdb script and detaches.
```
```suggestion
This shell script attaches GDB to the running dashd process (should be built with debug info), executes debug.gdb script and detaches.
```

PastaPastaPasta commented on contrib/auto_gdb/README.md:
```diff
@@ -0,0 +1,45 @@
+
+## dash_dbg.sh
+This shell script attaches GDB to running dashd process (should be built with debug info), executes debug.gdb script and detaches.
+By default it uses testnet setting,  see script comments to attach to mainnet dashd.
```
```suggestion
By default it uses testnet settings, see script comments to attach it to mainnet dashd.
```

PastaPastaPasta commented on contrib/auto_gdb/README.md:
```diff
@@ -0,0 +1,45 @@
+By default it uses testnet setting,  see script comments to attach to mainnet dashd.
+
+## debug.gdb
+Contains debugger instructions to execute during attach: load python code and executes it for the objects we want to investigate.
```
```suggestion
Contains debugger instructions to execute during attach: loads python code and executes it for the objects we want to investigate.
```

PastaPastaPasta commented on contrib/auto_gdb/README.md:
```diff
@@ -0,0 +1,45 @@
+Contains debugger instructions to execute during attach: load python code and executes it for the objects we want to investigate.
+
+## log_size.py
+Contains definition of the gdb command log_size. After loading of this script it can be called from gdb command line or other gdb scripts.
```
```suggestion
Contains definition of the gdb command log_size. After this script loads it can be called from gdb command line or other gdb scripts.
```

PastaPastaPasta commented on contrib/auto_gdb/README.md:
````diff
@@ -0,0 +1,45 @@
+```
+
+## stl_containers.py
+Contains helper classes to calculate memory used by the STL containers (list, vector, map, set, pair)
````
```suggestion
Contains helper classes to calculate memory used by the STL containers (list, vector, map, set, pair).
```

PastaPastaPasta commented on contrib/auto_gdb/README.md:
```diff
@@ -0,0 +1,45 @@
+Contains helper classes to calculate memory used by the STL containers (list, vector, map, set, pair)
+
+## simple_class_obj.py
+Contains helper class to calculate memory used by the object as a sum of the memory used by its fields.
```
```suggestion
Contains a helper class to calculate the memory used by an object as a sum of the memory used by its fields.
```

PastaPastaPasta commented on contrib/auto_gdb/README.md:
```diff
@@ -0,0 +1,45 @@
+
+## simple_class_obj.py
+Contains helper class to calculate memory used by the object as a sum of the memory used by its fields.
+All processed objects of such type are listed in the this file,  you can add new types you are interested in to this list.
```
```suggestion
All processed objects of such type are listed in the this file, you can add new types you are interested in to this list.
```

PastaPastaPasta commented on contrib/auto_gdb/README.md:
```diff
@@ -0,0 +1,45 @@
+## simple_class_obj.py
+Contains helper class to calculate memory used by the object as a sum of the memory used by its fields.
+All processed objects of such type are listed in the this file,  you can add new types you are interested in to this list.
+If some type is not listed here,  his size is calculated as sizeof result (except STL containers which are processed in stl_containers.py).
```
```suggestion
If a type is not listed here, its size is the return of sizeof (except STL containers which are processed in stl_containers.py).
```

gladcow commented:
Fixed
